File: androidApp/src/main/java/io/redlink/more/app/android/activities/observations/questionnaire/QuestionViewModel.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license with Commons Clause
 * (see https://www.apache.org/licenses/LICENSE-2.0 and
 * https://commonsclause.com/).
 */

package io.redlink.more.app.android.activities.observations.questionnaire

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.redlink.more.app.android.MoreApplication
import io.redlink.more.models.QuestionType
import io.redlink.more.viewModels.simpleQuestion.QuestionCoreViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class QuestionViewModel(
    scheduleId: String?,
    notificationId: String?,
    observationId: String?
) :
    ViewModel() {
    val coreViewModel: QuestionCoreViewModel = QuestionCoreViewModel(
        MoreApplication.shared!!.repositories,
        MoreApplication.shared!!.observationFactory,
        scheduleId,
        notificationId,
        observationId
    )

    val hasData = mutableStateOf(false)

    init {
        viewModelScope.launch(Dispatchers.Main.immediate) {
            coreViewModel.questionModel.collect { model ->
                withContext(Dispatchers.Main) {
                    hasData.value = model?.isValidModel() ?: false
                }
            }
        }
    }

    fun viewDidAppear() {
        coreViewModel.viewDidAppear()
    }

    fun viewDidDisappear() {
        coreViewModel.viewDidDisappear()
        hasData.value = false
    }

    fun finish(type: QuestionType, answer: Any) {
        coreViewModel.finishQuestion(answer)
    }
}

File: iosApp/iosApp/Views/QuestionObservation/QuestionViewModel.swift
//
//  SimplequestionObservationViewModel.swift
//  iosApp
//
//  Created by Isabella Aigner on 27.03.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung
//  Licensed under the Apache 2.0 license with Commons Clause
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//

import shared

import Combine
import KMPNativeCoroutinesCombine

class QuestionViewModel: ObservableObject {
    private let coreModel: QuestionCoreViewModel

    @Published var questionModel: QuestionModel?
    @Published var answers: [String] = []
    
    private var cancellables = Set<AnyCancellable>()

    init(navigationState: NavigationState) {
        coreModel = QuestionCoreViewModel(repository: AppDelegate.shared.repositories, observationFactory: AppDelegate.shared.observationFactory, scheduleId: navigationState.scheduleId, notificationId: navigationState.notificationId, observationId: navigationState.observationId)
        
        createPublisher(for: coreModel.questionModel)
            .receive(on: DispatchQueue.main)
            .sink(receiveCompletion: {_ in}) { [weak self] model in
                self?.questionModel = model
                self?.answers = (model?.answers as? [NSString])?.map { $0 as String } ?? []
            }
            .store(in: &cancellables)
    }

    func viewDidAppear() {
        coreModel.viewDidAppear()
    }

    func viewDidDisappear() {
        coreModel.viewDidDisappear()
    }
    
    func finish(data: AnyObject) {
        coreModel.finishQuestion(data: data)
    }
}

